docs/scripts/patch_devsite_toc.py

```diff
@@ -0,0 +1,173 @@
+        print(f"Adding section {section.title}...")
+        current_toc[0]["items"].insert(-1, section.to_dict())
+        section.copy_markdown()
+    # save file
+    if output_file_path is None:
+        output_file_path = toc_file_path
+    with open(output_file_path, "w") as f:
+        yaml.dump(current_toc, f)
+
+
+class TocSection:
+    def __init__(self, dir_name, index_file_name):
+        """
+        :param dir_name: name of the directory containing the rst files
+        :param index_file_name: name of an index file within dir_name. This file
+            will not be included in the table of contents, but provides an ordered
+            list of the other files which should be included
+        """
+        self.dir_name = dir_name
+        self.index_file_name = index_file_name
+        index_file_path = os.path.join(dir_name, index_file_name)
+        # find set of files referenced by the index file
+        with open(index_file_path, "r") as f:
+            self.title = f.readline().replace("\n", "")
+            in_toc = False
+            self.items = []
+            for line in f:
+                if line.startswith(".. toctree::"):
+                    in_toc = True
+                elif in_toc:
+                    if not line.strip():
+                        # ignore empty lines
+                        continue
+                    elif line.startswith(" ") or line.startswith("\t"):
+                        # keep track of the items in the table of contents
+                        # ignore : directives
+                        if ":" not in line:
+                            file_name = line.strip()
+                            # load the file to get the title
+                            with open(f"{dir_name}/{file_name}.rst", "r") as f2:
+                                file_title = f2.readline().replace("\n", "")
+                                self.items.append(
+                                    {"name": file_title, "href": f"{file_name}.md"}
+                                )
+                    else:
+                        # if we are out of the indented block, we are done
+                        break
+                        in_toc = False
+
+    def to_dict(self):
+        """
+        Convert the TocSection object to a dictionary that can be written to a yaml file
+        """
+        return {"name": self.title, "items": self.items}
+
+    def copy_markdown(self):
+        """
+        Copy markdown files from _build/markdown/dir_name to _build/html/docfx_yaml
+
+        This is necessary because the markdown files in dub-directories
+        are not copied over by the docfx build by default
+        """
+        for file in os.listdir("_build/markdown/" + self.dir_name):
+            shutil.copy(
+                f"_build/markdown/{self.dir_name}/{file}",
+                f"_build/html/docfx_yaml",
+            )
+
+
+def validate_toc(toc_file_path, expected_section_list, added_sections):
```

Comment:
at what point in the release process does this run? Can we end up in a state that the artifact was uploaded to pypi but this failed so we end up with a release egg but no docs?

If this is the case, should the validation happen as a presubmit instead?